### PR TITLE
Add the Directus Marketplace

### LINKED ISSUE #19800: Project Forum: Extensions Marketplace
Parent-issue to keep track of timelines and any additional subtasks for the marketplace 🚀 

```[tasklist]
### Must Have
- [ ] https://github.com/directus/directus/issues/19803
- [ ] https://github.com/directus/directus/issues/20625
- [ ] https://github.com/directus/directus/issues/20540
- [ ] https://github.com/directus/directus/issues/20559
- [ ] https://github.com/directus/directus/issues/20981
- [ ] https://github.com/directus/directus/issues/20539
- [ ] https://github.com/directus/directus/issues/20978
- [ ] https://github.com/directus/directus/issues/20560
- [ ] https://github.com/directus/directus/issues/20980
- [ ] https://github.com/directus/directus/issues/21005
- [ ] https://github.com/directus/directus/issues/21314
```
```[tasklist]
### Should Have
```

```[tasklist]
### Could Have
- [ ] Allow extensions to register their own global options
- [ ] Allow extensions to have an install/uninstall step (for database migrations / other prep)
- [ ] Env var / config to only allow packages from allowlisted / verified authors
- [ ] Upgrade / Downgrade Extensions
```

### Won't Have

- x

### LINKED ISSUE #21540: Env var for max_extensions

### PR DESCRIPTION
**Pull request by rijkvanzanten**
## Scope

What's changed:

- Added a new Marketplace section in settings that lets you browse and install extensions from our new Extensions Registry
- Added the ability to uninstall extensions from Settings
- Standardized the extensions folder by dropping the deprecated folder-per-type method
- Added a new migrations path setting to move migrations out of the extensions folder

## Potential Risks / Drawbacks

- It's mostly new code, so that's good. That being said..:
- This will allow you to download and run third party code which obviously comes with it's own risks

## Review Notes / Questions

- Please divert your gaze to https://github.com/orgs/directus/projects/26/views/1 for the up to date requirements list to track the status of the want-to-haves.
- Please note any and everything in the thread down below, and I'll put it on the requirements list.
- A lot of extensions currently present in the marketplace are outdated / deprecated. We should make sure we go through and mark the ones we know are good.

Closes #19800, closes #21540

### CODE REVIEW

**Review by br41nslug — COMMENTED** [review]
The reload page navigation guard and notification seem to be broken and forces you to refresh aggressively on each enable/disable without any way to cancel.

Reloading on `main`:

https://github.com/directus/directus/assets/9389634/94c0ee7c-d2a6-4c78-9119-19b0350b4810


Reloading on `next`:


https://github.com/directus/directus/assets/9389634/420a2771-685d-4b09-83f1-b011aedc253f


Having to do a full refresh then scroll back to the right extension to enable/disable the next one is not a great experience.

Pagination on the Marketplace has some room for improvement:
- navigating between pages is neither registered in the URL or in the browser history (so the browser back/forward dont work as expected)
- pagination does not reset when switching categories, if you go to page 10 on "all" extensions and then open "bundles" it will try to load page 10 which does not exist
- nice-to-have, a pagination button to jump directly to the first or last page as you cannot see all pages


https://github.com/directus/directus/assets/9389634/58195e46-76ce-4423-b9bb-a88cfe1c2cdf

**Review by DanielBiegler — CHANGES_REQUESTED** [review]
Search field rendering errors + clear-value icon needs some padding

[Screencast from 2024-02-07 14-38-44.webm](https://github.com/directus/directus/assets/14810858/9732fbd0-f9f9-4b16-86c9-ea52f16dae53)

**Review by DanielBiegler — CHANGES_REQUESTED** [review]
An indicator for empty lists would be nice. This is relevant for both

1. When there are no extensions for that extension type
2. When the user filters search results via the search-input and there are no results

| Currently | How about something similar to our empty v-form but of course with better wording? |
|--------|--------|
| image | image |

**Review by DanielBiegler — CHANGES_REQUESTED** [review]
Tiny tweaks:

1. Links inside the extension readme markdown should open in a new tab
2. Make said markdown text selectable. I wanted to look up some terminology used in a readme but was unable to select the text which felt bad from a ux standpoint

**Review by DanielBiegler — COMMENTED** [review]
Would be neat if we could jump from an extension to its marketplace entry so we can check out the readme again, for example

Maybe something similar to this:

image

**Review by azrikahar — COMMENTED** [review]
One tiny potential improvement is when an extension shows "Incompatible with your Project", what if we show the semver string as a tooltip on hover? Just as a way to be a tad more informative.

**Review by DanielBiegler — COMMENTED** [review]
Similar to @paescuj suggestion of filtering out empty/default READMEs I'd like to also ignore "banner images" for the tagline because I found a couple that will then show up like so which looks bad:

image

By "banner image" I mean repos similar to ours which features a banner as the first thing in the README, example:

image

**Review by DanielBiegler — COMMENTED** [review]
Might be good to only start the installation process after a confirmation modal to avoid accidental misclick-installing extensions

File: app/package.json [hunk 1]
```diff
@@ -140,5 +141,9 @@
 		"vue-tsc": "1.8.27",
 		"vuedraggable": "4.1.0",
 		"wellknown": "0.5.0"
+	},
+	"dependencies": {
+		"@directus/extensions-registry": "workspace:*",
+		"semver": "7.5.4"
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/package.json:149]: Good one! I keep forgetting everything is "Dev" for the app

File: packages/utils/node/list-folders.ts [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
 	const files = await fse.readdir(fullPath);
 
 	const directories: string[] = [];
 
 	for (const file of files) {
 		const filePath = path.join(fullPath, file);
+
+		if (options?.ignoreHidden && file.startsWith('.')) {
+			continue;
+		}
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply packages/utils/node/list-folders.ts:22]: Keeping me sharp, love it

**Comment by rijkvanzanten** [comment]
CleanShot 2024-02-09 at 16 41 23@2x

Upgraded sorting / search bar to resolve a bunch of the mentioned points 🙂

**Comment by rijkvanzanten** [comment]
OK all! I think we're ready for another round of reviews and merging of this PR 🙂 

There's a couple points I've jotted down as should/could haves for additional PRs that I don't believe are blocking this one, namely:

- Syncing filters/pagination with URL to maintain navigation[^1]
- Adding fallbacks for media / description / readmes
- Adding a link back to the npm page of a package [^2]
- Opening readme links in a new tab[^3]

[^1]: This is kinda a system wide issue on layouts 👀 
[^2]: Packages aren't always pulled from `npm`
[^3]: This is one of those things that can be surprisingly tricky as it's all rendered markdown rather than components.

**Comment by paescuj** [comment]
> 2. Packages aren't always pulled from `npm` [↩](#user-content-fnref-2-a3d2511aeb941f1a71b8922e7e9f6e48)

How about a general link to the package in the original registry, if available (be it npm, gh, ...)? Me and @phazonoverload both think it would be quite valuable to have such a link, and particularly also the original name of the package (since we format it differently), so that the origin of the package is clear.

> 3. This is one of those things that can be surprisingly tricky as it's all rendered markdown rather than components. [↩](#user-content-fnref-3-a3d2511aeb941f1a71b8922e7e9f6e48)

Thankfully, the `md` util got us covered 😃
``` 
v-md="{ value: readme, target: '_blank' }"
```

**Review by br41nslug — CHANGES_REQUESTED** [review]
Accessibility:
- navigating between pages is neither registered in the URL or in the browser history causing the browser buttons to not work correctly and have quite a frustrating experience while searching/browser extensions.
- What properties does search search through? Does this not include author? I am unable to find any of my extensions by either searching for `brainslug` or `br41nslug` while there are some if you manually go through the pages (i did allow non-sandbox) [In the short term it may be better not to have search]
- Relative URLs in README's are not correctly rendered with the github domain (see the test link https://www.npmjs.com/package/directus-extension-websocket-subscribe)

Design:
- [x] The type dropdown is not sizing properly with its scrollbar (optimally this would just show all items on my desktop where there is plenty of room not to need the scroll bar)
image
- [x] The search bar looks a tad funky in dark mode
- [ ] Since it can take some time for the interface to update (which i think will only get slower with more extensions) i think it would be good to have some form of loader state

https://github.com/directus/directus/assets/9389634/6fa4926d-314d-43ac-9848-c6170854b9c9


Errors:
- [x] Not exactly sure where this error is rendered but could we add a more friendly fallback here in case of no readme data found, npm gives us the error "This package does not have a README."
image


Extensions:
- [ ] Something is not playing well with my pre-existing extensions, not exactly sure what caused this but cleared the db table for further tests and that resolved it
image
- [ ] The extension page should have a badge to distinguish 2 extensions with the same name from different sources (`local` vs `registry`)
image
- [x] App extensions do not seem to load while the extension page says enabled, tested with modules and operations
image
image
image

File: app/src/modules/settings/routes/marketplace/routes/registry/components/inline-filter.vue [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+}>();
+
+const searchInputValue = ref('');
+
+watchDebounced(
+	searchInputValue,
+	(val) => {
+		search.value = val;
+	},
+	{ debounce: 300, maxWait: 1000 },
```
[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/modules/settings/routes/marketplace/routes/registry/components/inline-filter.vue:24]: I think that might've been the maxWait. Removed that, lets see if that solves it! 🤞🏻

**Comment by rijkvanzanten** [comment]
@br41nslug 

> What properties does search search through? Does this not include author? I am unable to find any of my extensions by either searching for brainslug or br41nslug while there are some if you manually go through the pages (i did allow non-sandbox) [In the short term it may be better not to have search]

I think this is one of the (classic) issues of Directus itself: The author is a UUID field on the parent package, and the search just uses `?search=` in the Directus API so doesn't search through author name 🤔

**Comment by rijkvanzanten** [comment]
> Ah that makes a lot of sense 🤔 should we perhaps do a more targeted filter including relational fields like author (not sure if there are any others that make sense to search through)

We could theoretically add a `?deep[author][name][_contains]=brainslug` but that would return less instead of more results 🤔

**Comment by rijkvanzanten** [comment]
OK I'm pens-down on this all! No more improvements on features will be added on this branch. The vast majority of improvement suggestions have been implemented. Lets switch our efforts for one more review to make sure everything works as expected and lets ship this sucker! 🚢

**Review by br41nslug — COMMENTED** [review]
I think the "Compatible with your Project" text is a bit misleading or creating false expectations. I've installed a bunch of the available extensions from the marketplace and this label seems counter intuitive:

I've installed "compatible" extensions that absolutely broke all extension loading

image
Code_PHxlY6qS0F

And vice versa installed "not verified to be compatible" extensions made for `9.X` that just work

image

The wording used "Compatible **with your Project**" and "Compatibility **not verified**" implies that we're doing more to actually verify extensions than just the simple string version check that's happening.

**Comment by rijkvanzanten** [comment]
@br41nslug Any suggestions? It's the way the extension author communicates whether or not they believe their extension is compatible with the project 🤔 There's no way for us to know if it actually is without trying them all

**Comment by br41nslug** [comment]
Changing the label to something like `Built for Directus version XXXX` or something along those lines. Or something that does not imply we're actually checking these extensions like the current wording.

**Review by br41nslug — COMMENTED** [review]
An old comment seems to have turned into a bug
> The extension page should have a badge to distinguish 2 extensions with the same name from different sources (local vs registry)

Where it used to render both installed versions (which was confusing but accurate):
Previous screenshot:
image

Current state:
image

Since we will have 3 different sources/methods of installing extensions (`local`, `npm` and `marketplace`) being able to spot and fix installations having multiple versions of the same extension installed does not feel like a luxury.

File: app/src/modules/settings/routes/marketplace/routes/registry/registry.vue [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+import RegistryInfoSidebarDetail from './components/registry-info-sidebar-detail.vue';
+
+const { t } = useI18n();
+
+const perPage = 10;
+
+const search = useRouteQuery('search');
+const page = useRouteQuery('page', 1);
+const type = useRouteQuery('type');
+const sort = useRouteQuery<'popular' | 'recent' | 'downloads'>('sort', 'popular');
```
Comment (DanielBiegler) [inline app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:21]: Having state in the URL is really nice now, but the initial problem of not being able to properly use the browsers back button remains.

1. Go to settings
2. Go to marketplace
3. Press onto some pagination page button, lets say page 3
4. Press the back button
5. You will now go back to the settings instead of going back to the first market place page

This is because we dont push the new URL to the page history

Comment (rijkvanzanten) [reply app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:21]: Curse and a blessing? 🤔 It would also be super annoying to have everything in the history state cause then you can never click back to get back to the previous page

Comment (br41nslug) [reply app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:21]: I look at it differently 😬 And do consider each page in "pagination" to be its own page (or more broadly each unique filter combination of `type`,`page`,`sort`)

> It would also be super annoying to have everything in the history state cause then you can never click back to get back to the previous page

In its current state it is impossible to get back to the previous page because it does not exist in the history while when everything is on the browser history it is a question of going back enough times to get where you wanted to be (it can take a while if you've done a lot of browsing but at least you can get back if wanted).

Comment (DanielBiegler) [reply app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:21]: FWIW: I'm in the same camp as @br41nslug - Personally I'd strongly prefer it being in the history :+1:

Comment (rijkvanzanten) [reply app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:21]: Can we push stuff in history in Vue without triggering a navigation event (eg reload)? If that's possible I'm willing to concede.

Comment (paescuj) [reply app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:21]: I think a direct [`history.pushState`](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState) call should do it.

Comment (rijkvanzanten) [reply app/src/modules/settings/routes/marketplace/routes/registry/registry.vue:21]: Spend about an hour or so trying to sync it into `history.pushState` directly, but am getting some very weird behavior. I'm assuming it's because of vue-router and the browser history getting out of sync while manually adjusting the browser history state. Using `vue-router` `route.push` directly doesn't seem to actually push it into the history state. I'm assuming it's because the vue-router is already on the 'current' route when the change happens. It won't add an entry if it's already the page we're on.

Long story short, I'm treating it as a nice to have but not a blocker for this PR. Lets plan on circling back to this later 👍🏻

**Comment by rijkvanzanten** [comment]
> Changing the label to something like Built for Directus version XXXX or Built for v10.4.3 or something along those lines. Im just suggesting we avoid the use of words like "compatible" or "verified" which people may read literally thinking the extension will actually work or was tested.

The goal for the host version is that extension authors can communicate what versions their extension has been tested on / confirmed working. If that value is incorrect for an individual extension, it should be updated on the extension's repo. On day 1, there will be a lot of extensions where this information is out of date, as we've never surfaced it before. I'm expecting that to smoothen out over time as the community develops more extensions

**Comment by br41nslug** [comment]
> The goal for the host version is that extension authors can communicate what versions their extension has been tested on / confirmed working.
> On day 1, there will be a lot of extensions where this information is out of date, as we've never surfaced it before.

I agree, but given that we expect this data to be incorrect lets be very clear here that it was the author of the extension that has "tested on / confirmed working" the extension not us Directus.

**Comment by rijkvanzanten** [comment]
Continued in https://github.com/directus/directus/pull/21674 due to some git shenanigans
